### PR TITLE
[wasm][debugger] View object attributes - DebugType=full

### PR DESCRIPTION
**Pull request by thaystg**
Fix view object attributes from a class defined in an assembly with debugType = full.

Fix discussion on: https://github.com/mono/mono/issues/19942

Adding the Runtime.consoleAPICalled, I couldn't use the Log.entryAdded didn't work on VS.

in VS:
image

in Chrome:
image

### CODE REVIEW

File: src/mono/wasm/debugger/tests/debugger-test/debugger-test.cs [hunk 1]
```diff
@@ -849,3 +849,16 @@ public static void RunDebuggerBreak()
...
+
+public class DebugTypeFull
+{
+    public static void CallToEvaluateLocal(string asm_base64, string pdb_base64)
+    {
+        var asm = System.Reflection.Assembly.LoadFrom("debugger-test-with-full-debug-type.dll");
+        var myType = asm.GetType("DebuggerTests.ClassToInspectWithDebugTypeFull");
+        var myMethod = myType.GetConstructor(new Type[] { });
+        var a = myMethod.Invoke(new object[]{});
+        Console.WriteLine(a);
```
Comment (ilonatommy) [inline src/mono/wasm/debugger/tests/debugger-test/debugger-test.cs:861]: Should we leave this writing to the console here?

Comment (thaystg) [reply src/mono/wasm/debugger/tests/debugger-test/debugger-test.cs:861]: Removed :)

File: src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs [hunk 2]
```diff
@@ -541,10 +541,19 @@ public unsafe AssemblyInfo(string url, byte[] assembly, byte[] pdb)
...
             {
                 pdbStream = new MemoryStream(pdb);
-                pdbMetadataReader = MetadataReaderProvider.FromPortablePdbStream(pdbStream).GetMetadataReader();
+                try
+                {
+                    pdbMetadataReader = MetadataReaderProvider.FromPortablePdbStream(pdbStream).GetMetadataReader();
+                }
+                catch (BadImageFormatException)
+                {
+                    Console.WriteLine($"Warning: Unable to read debug information of: {Name} (use DebugType=Portable/Embedded)");
```
Comment (lewing) [inline src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs:555]: you could log this using https://chromedevtools.github.io/devtools-protocol/tot/Log/ have it surface to the frontend

File: src/mono/wasm/debugger/tests/debugger-test/debugger-test.csproj [hunk 3]
```diff
@@ -22,6 +22,7 @@
     <ProjectReference Include="..\library-dependency-debugger-test2\library-dependency-debugger-test2.csproj" Private="true"/>
     <ProjectReference Include="..\debugger-test-with-source-link\debugger-test-with-source-link.csproj" Private="true"/>
     <ProjectReference Include="..\ApplyUpdateReferencedAssembly\ApplyUpdateReferencedAssembly.csproj" />
+    <ProjectReference Include="..\debugger-test-with-full-debug-type\debugger-test-with-full-debug-type.csproj" Private="true"/>
```
Comment (lewing) [inline src/mono/wasm/debugger/tests/debugger-test/debugger-test.csproj:25]: does this need to be added?

File: src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs (hunk 2 above)
Comment (lewing) [reply src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs:555]: You emit https://chromedevtools.github.io/devtools-protocol/tot/Log/#event-entryAdded from the proxy?

File: src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs [hunk 4]
```diff
@@ -541,10 +541,19 @@ public unsafe AssemblyInfo(string url, byte[] assembly, byte[] pdb)
...
             asmMetadataReader = PEReaderExtensions.GetMetadataReader(peReader);
+            Name = asmMetadataReader.GetAssemblyDefinition().GetAssemblyName().Name + ".dll";
+            AssemblyNameUnqualified = asmMetadataReader.GetAssemblyDefinition().GetAssemblyName().Name + ".dll";
             if (pdb != null)
             {
                 pdbStream = new MemoryStream(pdb);
-                pdbMetadataReader = MetadataReaderProvider.FromPortablePdbStream(pdbStream).GetMetadataReader();
+                try
+                {
+                    pdbMetadataReader = MetadataReaderProvider.FromPortablePdbStream(pdbStream).GetMetadataReader();
```
Comment (radical) [inline src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs:551]: which method throws `BadImageFormatException`?

File: src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs [hunk 5]
```diff
@@ -541,10 +541,19 @@ public unsafe AssemblyInfo(string url, byte[] assembly, byte[] pdb)
             asmStream = new MemoryStream(assembly);
             peReader = new PEReader(asmStream);
             asmMetadataReader = PEReaderExtensions.GetMetadataReader(peReader);
+            Name = asmMetadataReader.GetAssemblyDefinition().GetAssemblyName().Name + ".dll";
+            AssemblyNameUnqualified = asmMetadataReader.GetAssemblyDefinition().GetAssemblyName().Name + ".dll";
             if (pdb != null)
             {
                 pdbStream = new MemoryStream(pdb);
```
Comment (radical) [inline src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs:548]: IIUC, `pdbStream` doesn't really ever get used again. We instantiate it here, and use it to get the `pdbMetadataReader`. And in other methods, we re-instantiate.
And same for `asmReader`.
If so, then we can remove the fields, and dispose these.

File: src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs [hunk 6]
```diff
@@ -541,10 +541,19 @@ public unsafe AssemblyInfo(string url, byte[] assembly, byte[] pdb)
             asmStream = new MemoryStream(assembly);
             peReader = new PEReader(asmStream);
             asmMetadataReader = PEReaderExtensions.GetMetadataReader(peReader);
+            Name = asmMetadataReader.GetAssemblyDefinition().GetAssemblyName().Name + ".dll";
+            AssemblyNameUnqualified = asmMetadataReader.GetAssemblyDefinition().GetAssemblyName().Name + ".dll";
```
Comment (radical) [inline src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs:545]: Looking at the implementation, `GetAssemblyDefinition()` returns a new object on every call. And `GetAssemblyName` constructing a new `AssemblyName` instance. so, we should avoid doing that twice.

File: src/mono/wasm/debugger/tests/debugger-test/debugger-test.csproj (hunk 3 above)
Comment (thaystg) [reply src/mono/wasm/debugger/tests/debugger-test/debugger-test.csproj:25]: I think so, to compile the new project for the test case.

File: src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs (hunk 4 above)
Comment (thaystg) [reply src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs:551]: `FromPortablePdbStream`

File: src/mono/wasm/debugger/tests/debugger-test/debugger-test.csproj (hunk 3 above)
Comment (lewing) [reply src/mono/wasm/debugger/tests/debugger-test/debugger-test.csproj:25]: right, I mean did you forget to add the project to the pr?

Comment (thaystg) [reply src/mono/wasm/debugger/tests/debugger-test/debugger-test.csproj:25]: I'm sorry Larry!

**Review by radical — DISMISSED** [review]

File: src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs [hunk 7]
```diff
@@ -526,37 +526,49 @@ internal class AssemblyInfo
...
+            var asmStream = new MemoryStream(assembly);
+            var peReader = new PEReader(asmStream);
+            var entries = peReader.ReadDebugDirectory();
+            var codeView = entries[0];
+            var codeViewData = peReader.ReadCodeViewDebugDirectoryData(codeView);
+            PdbAge = codeViewData.Age;
+            PdbGuid = codeViewData.Guid;
+            PdbName = codeViewData.Path;
             asmMetadataReader = PEReaderExtensions.GetMetadataReader(peReader);
+            var asmDef = asmMetadataReader.GetAssemblyDefinition();
```
[the remark this answers is not in the document]
Comment (thaystg) [reply src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs:555]: removed the `AssemblyNameUnqualified `as it has the same value of `Name`

File: src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs (hunk 2 above)
Comment (thaystg) [reply src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs:567]: As far as I understood this is not supported by JSDebugger so I implemented using `Runtime.consoleAPICalled`.

File: src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs [hunk 8]
```diff
@@ -526,37 +526,49 @@ internal class AssemblyInfo
...
 
         public unsafe AssemblyInfo(string url, byte[] assembly, byte[] pdb)
         {
             this.id = Interlocked.Increment(ref next_id);
-            asmStream = new MemoryStream(assembly);
-            peReader = new PEReader(asmStream);
+            var asmStream = new MemoryStream(assembly);
+            var peReader = new PEReader(asmStream);
+            var entries = peReader.ReadDebugDirectory();
+            var codeView = entries[0];
```
[the remark this answers is not in the document]
Comment (thaystg) [reply src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs]: done.

File: src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs [hunk 9]
```diff
@@ -526,37 +526,49 @@ internal class AssemblyInfo
...
+            var codeViewData = peReader.ReadCodeViewDebugDirectoryData(codeView);
+            PdbAge = codeViewData.Age;
+            PdbGuid = codeViewData.Guid;
+            PdbName = codeViewData.Path;
             asmMetadataReader = PEReaderExtensions.GetMetadataReader(peReader);
+            var asmDef = asmMetadataReader.GetAssemblyDefinition();
+            Name = asmDef.GetAssemblyName().Name + ".dll";
+            AssemblyNameUnqualified = asmDef.GetAssemblyName().Name + ".dll";
```
[the remark this answers is not in the document]
Comment (thaystg) [reply src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs]: removed this property. as far as I undesrtood it has the same content of the `Name` property.

**Review by radical — APPROVED** [review]

**Pull request merged**